src/Symfony/Component/Security/Core/Authentication/ContainerAwareAuthenticationProviderManager.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authentication;
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Uses the Container to lazily instantiate the providers.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ContainerAwareAuthenticationProviderManager extends AbstractAuthenticationProviderManager
+{
+    private $providerServiceIds;
+
+    private $container;
+
+    private $eraseCredentials;
+
+    public function __construct(array $providerServiceIds, ContainerInterface $container, $eraseCredentials = true)
+    {
+        $this->providerServiceIds = $providerServiceIds;
+        $this->container = $container;
+        $this->eraseCredentials = $eraseCredentials;
+    }
+
+    protected function getProviders()
+    {
+        $providers = array();
+        foreach ($this->providerServiceIds as $serviceId) {
+            $providers[] = $this->container->get($serviceId);
+        }
```

Comment:
Maybe this list should be cached internally if getProviders can be called more often. The other option would be to yield the providers. This solution would mean that the earlier it finds a provider, the less get calls to the container are executed (if I understand yield properly that is).